src/libopensc/card-skeid.c

```diff
@@ -0,0 +1,180 @@
+	u8 buf[SKEID_KNOWN_URL_LEN];
+
+	sc_path_t url_path;
+
+	int r = SC_ERROR_WRONG_CARD;
+
+	sc_path_set(&url_path, SC_PATH_TYPE_DF_NAME, skeid_aid_eid.value, skeid_aid_eid.len, 0, 0);
+
+	if (sc_select_file(card, &url_path, NULL) == SC_SUCCESS
+		&&  sc_get_data(card, 0x7F62, buf, SKEID_KNOWN_URL_LEN) == SKEID_KNOWN_URL_LEN
+		&& !memcmp(buf, known_url, SKEID_KNOWN_URL_LEN))
+		r = SC_SUCCESS;
+
+	return r;
+}
+
+static int skeid_match_card(sc_card_t *card)
+{
+	if (_sc_match_atr(card, skeid_atrs, &card->type) < 0 || skeid_known_url(card) != SC_SUCCESS)
+		return 0;
+
+	sc_log(card->ctx,  "Slovak eID card v3 (CardOS 5.4)");
+
+	return 1;
+}
+
+static int skeid_get_serialnr(sc_card_t *card)
+{
+	int r;
+	sc_apdu_t apdu;
+	u8  rbuf[SC_MAX_APDU_BUFFER_SIZE];
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xca, 0x01, 0x81);
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = 256;
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r,  "APDU transmit failed");
+	if (apdu.sw1 != 0x90 || apdu.sw2 != 0x00)
+		return SC_ERROR_INTERNAL;
+	if (apdu.resplen == 8) {
+		/* cache serial number */
+		memcpy(card->serialnr.value, rbuf, 8);
+		card->serialnr.len = 8;
+	} else {
+		sc_log(card->ctx,  "unexpected response to GET DATA serial"
+			" number\n");
+		return SC_ERROR_INTERNAL;
+	}
+	return SC_SUCCESS;
+}
+
+static int skeid_init(sc_card_t *card)
+{
+	const unsigned long flags = SC_ALGORITHM_RSA_PAD_PKCS1 | SC_ALGORITHM_RSA_HASH_NONE;
+	const size_t data_field_length = 437;
+	int r;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	card->name = "Slovak eID (CardOS)";
+	card->type = SC_CARD_TYPE_SKEID_V3;
+	card->cla = 0x00;
+
+	r = skeid_get_serialnr(card);
+	LOG_TEST_RET(card->ctx, r, "Error reading serial number.");
+
+	card->caps |= SC_CARD_CAP_APDU_EXT | SC_CARD_CAP_ISO7816_PIN_INFO;
+
+	card->max_send_size = data_field_length - 6;
+#ifdef _WIN32
+	/* see card-cardos.c */
+	if (card->reader->max_send_size == 255 && card->reader->max_recv_size == 256) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "resetting reader to use data_field_length");
+		card->reader->max_send_size = data_field_length - 6;
+		card->reader->max_recv_size = data_field_length - 3;
+	}
+#endif
+
+	card->max_send_size = sc_get_max_send_size(card); /* see card-cardos.c */
+	card->max_recv_size = data_field_length - 2;
+	card->max_recv_size = sc_get_max_recv_size(card);
+
+	r = _sc_card_add_rsa_alg(card, 3072, flags, 0);
+
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int skeid_set_security_env(sc_card_t *card,
+		const sc_security_env_t *env,
+		int se_num)
+{
+	int key_id;
+	int r;
+
+	assert(card != NULL && env != NULL);
+
+	if (!(env->flags & SC_SEC_ENV_KEY_REF_PRESENT) || env->key_ref_len != 1) {
+		sc_log(card->ctx, "No or invalid key reference\n");
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+
+	key_id = env->key_ref[0];
+	r = sc_restore_security_env(card, key_id);
```

Comment:
This is at least weird. I would be probably for reimplementing the `iso7816_restore_security_env` (its just one APDU) with some comments, instead of reusing it for setting the security environment (with the key id instead of security environment number). The restore of security environment is completely different operation and that it works is probably pure coincidence.

Reply:
If others dont have issue with this. I am ok with keeping this as long as it will contain some comment in the code that this is what the official driver does. Without comment it still looks like an error (and comments on github are not guaranteed to be accessible forever).